### PR TITLE
algorithm: ZFunction

### PR DESCRIPTION
**Pull request by Adrito-M**
[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm?
- [ ] Documentation change?

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new JavaScript files are placed inside an existing directory.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
- [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/JavaScript/pull/1239"></a>

### CODE REVIEW

**Comment by Adrito-M** [comment]
@raklaptudirm @appgurueu  Please Take A Look

**Comment by appgurueu** [comment]
You don't need to ping us. Our review is automatically requested as we are code owners.

File: String/ZFunction.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * Author: Adrito Mukherjee
+ * Implementation of ZFunction in JavaScript
+ * ZFunction at an index i gives the length of the longest substring starting at i, that is also a prefix of the whole string
+ * ZFunction for all indices in a string can be calculated in O(N)
+ * Explanation: https://cp-algorithms.com/string/z-function.html
+ */
+
+function ZFunction (text) {
+  const length = text.length
+  const ZArray = Array(length).fill(0)
```
Comment (appgurueu) [inline String/ZFunction.js:11]: violating naming conventions

File: String/ZFunction.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/**
+ * Author: Adrito Mukherjee
+ * Implementation of ZFunction in JavaScript
+ * ZFunction at an index i gives the length of the longest substring starting at i, that is also a prefix of the whole string
+ * ZFunction for all indices in a string can be calculated in O(N)
+ * Explanation: https://cp-algorithms.com/string/z-function.html
+ */
+
+function ZFunction (text) {
```
Comment (appgurueu) [inline String/ZFunction.js:9]: also violating naming conventions

File: String/test/ZFunction.test.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import ZFunction from '../ZFunction'
+
+test('Test Case 1', () => {
```
Comment (appgurueu) [inline String/test/ZFunction.test.js:3]: Useless test case naming and useless `text` & `ZArray` variables.

**Review by appgurueu — CHANGES_REQUESTED** [review]
Please also add comments *in the code*.

File: String/test/ZFunction.test.js (hunk 3 above)
Comment (Adrito-M) [reply String/test/ZFunction.test.js]: Improved naming and removed useless variables

File: String/ZFunction.js (hunk 2 above)
Comment (Adrito-M) [reply String/ZFunction.js]: Corrected to `zFunction`

File: String/ZFunction.js (hunk 1 above)
Comment (Adrito-M) [reply String/ZFunction.js]: Changed to `zArray`

File: String/ZFunction.js [hunk 4]
```diff
@@ -1,32 +1,41 @@
 /**
- * Author: Adrito Mukherjee
+ * @author: Adrito Mukherjee
  * Implementation of ZFunction in JavaScript
  * ZFunction at an index i gives the length of the longest substring starting at i, that is also a prefix of the whole string
  * ZFunction for all indices in a string can be calculated in O(N)
  * Explanation: https://cp-algorithms.com/string/z-function.html
```
Comment (appgurueu) [inline String/ZFunction.js:6]: `@see`

File: String/test/ZFunction.test.js [hunk 5]
```diff
@@ -1,15 +1,10 @@
...
-  expect(ZArray).toEqual([10, 1, 0, 0, 2, 1, 0, 3, 1, 0])
-})
-
-test('Test Case 2', () => {
-  const text = 'aabxaabxcaabxaabxay'
-  const ZArray = ZFunction(text)
-  expect(ZArray).toEqual([
-    19, 1, 0, 0, 4, 1, 0, 0, 0, 8, 1, 0, 0, 5, 1, 0, 0, 1, 0
-  ])
+describe('Testing the zFunction', () => {
```
Comment (appgurueu) [inline String/test/ZFunction.test.js:3]: Just use one `test` block

File: String/ZFunction.js (hunk 4 above)
Comment (Adrito-M) [reply String/ZFunction.js]: Added `@see`

**Comment by Adrito-M** [comment]
Can you please add the `Hacktoberfest Accepted` Tag in case the PR isn't fit to be merged yet

**Review by appgurueu — APPROVED** [review]

**Review by raklaptudirm — APPROVED** [review]

**Pull request merged**